### PR TITLE
vendor docker/distribution d06d6d3b093302c02a93153ac7b06ebc0ffd1793

### PR DESCRIPTION
**Pull request by runcom**
Fix https://github.com/docker/docker/issues/21147
Partly address some test cases (external token service) for https://github.com/docker/docker/issues/21054

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by runcom** [comment]
ping @aaronlehmann

**Comment by aaronlehmann** [comment]
@runcom: https://github.com/docker/distribution/pull/1531 was just merged, so you can feel free to revendor `distribution` as part of this PR if it saves a step.

**Comment by runcom** [comment]
@aaronlehmann cool, running hack/vendor then

File: integration-cli/docker_cli_push_test.go [hunk 1]
```diff
@@ -577,4 +577,6 @@ func (s *DockerRegistryAuthTokenSuite) TestPushMisconfiguredTokenServiceResponse
 	dockerCmd(c, "tag", "busybox", repoName)
 	out, _, err := dockerCmdWithError("push", repoName)
 	c.Assert(err, check.NotNil, check.Commentf(out))
+	split := strings.Split(out, "\n")
+	c.Assert(split[len(split)-2], check.Equals, "unauthorized: authentication required")
```
Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: @aaronlehmann funny enough, now a bad formatted error won't result always in an unexpected http error but with the statuscode it translates to the correct error message in the client, is it the expected behavior?

Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: for instance, the token service in the test correctly returns 401 but fails at formatting the body - either case, the client get the correct error in the cli

Comment (aaronlehmann) [inline integration-cli/docker_cli_push_test.go]: @runcom: I believe there is a special case for 401 that returns the correct error. For other status codes, you should get "unexpected HTTP response body".

Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: alright, but before this distribution vendor I got just <nil> so 401 wasn't handled that way - I guess it's fine though right?

Comment (aaronlehmann) [inline integration-cli/docker_cli_push_test.go]: @runcom: Makes sense - looks like the special case wouldn't be hit with the old code. https://github.com/docker/distribution/pull/1531/files#diff-de05143a5d591b472e57dd8c46f3998fR88

Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: Alright, so besides tests this looks good

Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: Now that I think mode, I'll write a test with a 500 status code to make sure the cli prints the expected output

**Comment by runcom** [comment]
@aaronlehmann I added some other tests, but I cannot get the unexpected error string, seems like every status code is already handled in its own way. No big deal, I guess it's still fine.

**Comment by aaronlehmann** [comment]
@runcom: The response has to be unparseable JSON to trigger that message.

**Comment by runcom** [comment]
@aaronlehmann I meant I can't trigger `ErrNoErrorsInBody`

**Comment by aaronlehmann** [comment]
@runcom: Took another look at the code - needs to be a 4xx error other than 401.

**Comment by runcom** [comment]
@aaronlehmann thx, should be good now

**Comment by runcom** [comment]
```
PASS: docker_cli_push_test.go:583: DockerRegistryAuthTokenSuite.TestPushMisconfiguredTokenServiceResponseError	50.701s
PASS: docker_cli_push_test.go:595: DockerRegistryAuthTokenSuite.TestPushMisconfiguredTokenServiceResponseNoToken	50.726s
```
these tests pass but they are clearly *Retrying...*
Should a request ends up with xfer.DoNotRetry when it receives a 500 internal server error? (test 1)
Should a request ends up with  xfer.DoNotRetry when it does not contain any token? (test 2)

**Comment by aaronlehmann** [comment]
> Should a request ends up with xfer.DoNotRetry when it receives a 500 internal server error? (test 1)

I don't think so, because 500 Internal Server Error is often a temporary condition.

> Should a request ends up with xfer.DoNotRetry when it does not contain any token? (test 2)

That sounds reasonable. What do you think?

**Comment by runcom** [comment]
> Should a request ends up with xfer.DoNotRetry when it does not contain any token? (test 2)
> That sounds reasonable. What do you think?

If you're getting 200 and no token probably the registry is misconfigured and I don't see any reason in retrying, make sense?

**Comment by runcom** [comment]
@aaronlehmann this PR includes a change which need https://github.com/docker/distribution/issues/1542 merged and vendored (if it looks good ofc)

**Comment by thaJeztah** [comment]
Build is failing;

```
11:22:16 ---> Making bundle: binary (in bundles/1.11.0-dev/binary)
11:22:16 Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
11:22:33 # github.com/docker/docker/distribution
11:22:33 distribution/errors.go:97: undefined: auth.ErrNoToken
11:23:23 Build step 'Execute shell' marked build as failure
```

**Comment by runcom** [comment]
@thaJeztah https://github.com/docker/docker/pull/21283#issuecomment-198263665

**Comment by aaronlehmann** [comment]
@runcom: docker/distribution#1542 is merged. Thanks for the PR.

**Comment by runcom** [comment]
@aaronlehmann I'm updating this PR

**Comment by aaronlehmann** [comment]
@runcom: Looks like something went wrong with vendoring:

```
17:54:59 vendor/src/github.com/aws/aws-sdk-go/aws/ec2metadata/service.go:11:2: cannot find package "github.com/aws/aws-sdk-go/aws/service" in any of:
...
```

**Comment by runcom** [comment]
Yup, I'm afk right now, will figure it out once back home, though it seems aws deps are screwed

**Comment by runcom** [comment]
Fixed (somehow)

**Comment by calavera** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**